plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -648,6 +650,20 @@ evalBuiltinApp fun term runtime = case runtime of
     _ -> pure $ VBuiltin fun term runtime
 {-# INLINE evalBuiltinApp #-}
 
+data ArgsR uni fun ann
```

Comment:
So where is the speedup coming from? We can't unpack `CekValue` or `ArgsR`, so the only thing I can see is that it's spine-strict. Is that what you're expecting makes the difference?

If so, is there an advantage to having this fully-specialized one, rather than just defining a standalone spine-strict list type?

Reply:
Sorry, it's fully strict, not just spine-strict.

Reply:
And indeed, there are many strict lists defined elsewhere we could take. Unsure if it's worth it.

Reply:
> And indeed, there are many strict lists defined elsewhere we could take. 

I actually couldn't find any. There's `strict-list`, but I'm skeptical about it's alignment (it claims reversing things all the time is faster, but I failed to find any evidence for that) and it doesn't have enough `INLINE` pragmas.

> We can't unpack `CekValue` or `ArgsR`, so the only thing I can see is that it's spine-strict. 

I think Kmett told us that monomorphizing data definitions is beneficial even when nothing can be unpacked? I don't actually know and would like to learn. Defining `ArgsR` manually is extremely cheap though, just a few lines of very local code and it has the advantage of clearly conveying the fact that the args are in reverse order.

Reply:
I should've defined it as

```haskell
data ArgsR uni fun ann
    = NilArgsR
    | SnocArgsR !(ArgsR uni fun ann) !(CekValue uni fun ann)
```

BTW.

Reply:
He did? I don't remember that... but you might be right.

> Defining ArgsR manually is extremely cheap though, just a few lines of very local code and it has the advantage of clearly conveying the fact that the args are in reverse order.

IDK, having custom datatypes makes the reader load more concepts into their brains. It feels easier for me to just see "this is a strict list". Not sure.

Reply:
> He did? I don't remember that...

Or I made it up, which is very possible. I wonder where I can learn about this.

> IDK, having custom datatypes makes the reader load more concepts into their brains. It feels easier for me to just see "this is a strict list". Not sure.

I don't have any strong feelings about that anyway.